src/utils/notification/mail.ts

```diff
@@ -0,0 +1,105 @@
+import { sgSendMail } from '../mail'
+import { ConfigVariable, config } from '../config'
+import { EntityManager } from 'typeorm'
+import {
+  Account,
+  EmailDeliveryStatus,
+  Notification,
+  NextEntityId,
+  NotificationEmailDelivery,
+  OnChainNotificationEmailDelivery,
+} from '../../model'
+import { getNextIdForEntity } from '../nextEntityId'
+import { notificationEmailContent } from '../../auth-server/emails'
+import { linkForNotification } from './notificationLinks'
+import { getMemberHandle, textForNotification } from './notificationTexts'
+import { Flat } from 'lodash'
+import { EntityManagerOverlay } from '../overlay'
+
+// mail notification are immediately deposited into orion_db
+// FIXME: there could be a problem if the processor crashes and the overlay doesn't flush and also this can decrease performance
+export async function deliverNotificationViaEmail(
+  em: EntityManager,
+  toAccount: Account,
+  notification: Flat<Notification>
+): Promise<void> {
+  const nextEntityId = await getNextIdForEntity(em, 'OffChainNotificationEmailDelivery')
+  const notificationDelivery = new NotificationEmailDelivery({
+    id: nextEntityId.toString(),
+    notificationId: notification.id,
+    deliveryAttemptAt: new Date(),
+    deliveryStatus: EmailDeliveryStatus.Unsent,
+  })
+
+  const appName = await config.get(ConfigVariable.AppName, em)
+  const content = await createMailContent(em, toAccount, appName, notification)
+  notificationDelivery.deliveryStatus = await executeMailDelivery(appName, em, toAccount, content)
+  await em.save([
+    new NextEntityId({
+      entityName: 'OffChainNotificationEmailDelivery',
+      nextId: nextEntityId + 1,
+    }),
+    notificationDelivery,
+  ])
+}
+
+export async function deliverOnChainNotificationViaEmail(
+  store: EntityManagerOverlay,
+  toAccount: Account,
+  notification: Flat<Notification>
+): Promise<void> {
+  const id = store.getRepository(OnChainNotificationEmailDelivery).getNewEntityId()
+  const notificationDelivery = store.getRepository(OnChainNotificationEmailDelivery).new({
+    id,
+    notificationId: notification.id,
+    deliveryAttemptAt: new Date(),
+    deliveryStatus: EmailDeliveryStatus.Unsent,
+  })
+
+  const appName = await config.get(ConfigVariable.AppName, store.getEm())
+  const content = await createMailContent(store.getEm(), toAccount, appName, notification)
+  notificationDelivery.deliveryStatus = await executeMailDelivery(
+    appName,
+    store.getEm(),
+    toAccount,
+    content
+  )
```

Comment:
the on-chain notification delivery logic is still using the `em` instead of `overlay`, in `executeMailDelivery` function, so the delivery would fail if the cached content is not saved into the db

e.g here is the error in case of channel created event

<img width="714" alt="image" src="https://github.com/Joystream/orion/assets/37098720/cc66a846-eb46-4850-8115-6a9fdca7d164">

Reply:
The problem is not in the delivery, is in the content creation (as the error suggest)
It is trying to fetch the entity data (for some notification for which entity data is required) in order to construct the content from `orion_db` but the `overlay` hasn't flushed yet

Reply:
I need to fix the link and the text construction part